### PR TITLE
[HW] Adding support for Fixed-Point vector instructions

### PR DESCRIPTION
**Pull request by hossein1387**
Description of PR that completes issue here...

## Changelog
- Support for Fixed-Point vector instructions: `vaadd, vaaddu, vsadd, vsaddu, vssub, vssubu, vasub, vasubu`
- Two status register for fixed-point unit: `vxrm, vxsat`

### Fixed

- Description of changes

### Added

- Support for Fixed-Point vector instructions: `vaadd, vaaddu, vsadd, vsaddu, vssub, vssubu, vasub, vasubu`
- Two status register for fixed-point unit: `vxrm, vxsat`

## Checklist

- [x] Automated tests pass
- [x] No frequency degradation
- [x] Changelog updated
- [x] Code style guideline is observed

Please check our [contributing guidelines](CONTRIBUTING.md) before opening a Pull Request.

### CODE REVIEW

**Comment by mp-17** [comment]
Hey Hossein, thanks for this important contribution and sorry for the delay in the answer ;)
I see that the `riscv-tests` are not completely successful, are you still working on the PR?

**Comment by mp-17** [comment]
Hello Hossein, thanks you for the PR! 
I have still not checked precisely the assignments within the `simd_alu` in terms of math correctness, but I started a review process and you can address those points in the meantime.

BTW: can you please restart the CI? You can do that from the "Action" tab, or you can just push another commit (you will need to do it anyway to address some of the review points). Thanks a lot!

**Comment by hossein1387** [comment]
> Hello Hossein, thank you for the PR! I have still not checked precisely the assignments within the `simd_alu` in terms of math correctness, but I started a review process and you can address those points in the meantime.
> 
> BTW: can you please restart the CI? You can do that from the "Action" tab, or you can just push another commit (you will need to do it anyway to address some of the review points). Thanks a lot!

Hi Matteo, I am not able to see the restart option in the action tab. I am not sure if it is related to the privilege level or not. I can re submit the PR.

**Comment by mp-17** [comment]
Ehy Hossein, do you have any news? ;)

**Review by mp-17 — CHANGES_REQUESTED** [review]
Thanks for the PR, I only have some doubts and comments.

File: CHANGELOG.md
Comment (suehtamacv) [inline CHANGELOG.md:41]: Please use the present tense (`Add fixed-point instructions`). I also think this would be better suited for the `Added` section, instead of `Changed` section.

File: apps/riscv-tests/isa/macros/vector/vector_macros.h
Comment (suehtamacv) [inline apps/riscv-tests/isa/macros/vector/vector_macros.h:66]: Vertical alignment.

File: hardware/include/ara_pkg.sv
Comment (suehtamacv) [inline hardware/include/ara_pkg.sv:103]: Check whether we still have enough bits to encode all those opcodes.. I think Verilator would've issued a warning otherwise, but it does not hurt to check.

File: hardware/src/ara.sv
Comment (suehtamacv) [inline hardware/src/ara.sv:115]: Unnecessary whitespace changes.

File: hardware/src/lane/simd_alu.sv
Comment (suehtamacv) [inline hardware/src/lane/simd_alu.sv:260]: Are those changes needed?

**Review by suehtamacv — CHANGES_REQUESTED** [review]
Thanks for the PR, @hossein1387! I have a few comments that I would ask you to solve before we can merge this, but I think it looks good :)

**Comment by suehtamacv** [comment]
Well actually the CI is failing on `vsadd` with an Ara configuration which has two lanes.. That is something we should investigate

**Comment by M-Ijaz-10x** [comment]
Hey @hossein1387 , @suehtamacv 
We have been working on fixing the RTL failures of `vasub`, `vasubu` and `vssub` vfixed-point instructions, fixing CI check failures of `vsadd`, `vsaddu` and `vssub` instructions, and adding support for `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu` vfixed-point instructions. What we have done so far is as follows:
- We have added support for `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu` vfixed-point instruction and have tested them by using the provided tests
- We have fixed the RTL failures of `vasub`, `vasubu` and `vssub` vfixed-point instructions
- We have fixed CI check failures of `vsadd`, `vsaddu` and `vssub` instructions
- We have fixed `vssub.c` test which caused the spike failure

The final testing of all the bug-fixes and RTL additions is on its way. We are planning to push the fixes as soon as we get the final results.

Thanks

File: apps/riscv-tests/isa/rv64uv/Makefrag
Comment (hossein1387) [reply apps/riscv-tests/isa/rv64uv/Makefrag:12]: yes I do, I missed adding them into Makefrag

File: apps/riscv-tests/isa/macros/vector/vector_macros.h
Comment (hossein1387) [reply apps/riscv-tests/isa/macros/vector/vector_macros.h:46]: yes

File: hardware/src/lane/simd_alu.sv
Comment (hossein1387) [reply hardware/src/lane/simd_alu.sv:136]: it should default to `0` it seems I missed it.

Comment (hossein1387) [reply hardware/src/lane/simd_alu.sv:136]: yes we can, I was not sure if masking csr flags follows the ISA.

File: hardware/include/ara_pkg.sv
Comment (hossein1387) [reply hardware/include/ara_pkg.sv:103]: 108 of available 128 opcodes is used.

**Comment by hossein1387** [comment]
> Hey @hossein1387 , @suehtamacv We have been working on fixing the RTL failures of `vasub`, `vasubu` and `vssub` vfixed-point instructions, fixing CI check failures of `vsadd`, `vsaddu` and `vssub` instructions, and adding support for `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu` vfixed-point instructions. What we have done so far is as follows:
> 
> * We have added support for `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu` vfixed-point instruction and have tested them by using the provided tests
> * We have fixed the RTL failures of `vasub`, `vasubu` and `vssub` vfixed-point instructions
> * We have fixed CI check failures of `vsadd`, `vsaddu` and `vssub` instructions
> * We have fixed `vssub.c` test which caused the spike failure
> 
> The final testing of all the bug-fixes and RTL additions is on its way. We are planning to push the fixes as soon as we get the final results.
> 
> Thanks

That is great! is your changes based of off this PR?

**Comment by M-Ijaz-10x** [comment]
> > Hey @hossein1387 , @suehtamacv We have been working on fixing the RTL failures of `vasub`, `vasubu` and `vssub` vfixed-point instructions, fixing CI check failures of `vsadd`, `vsaddu` and `vssub` instructions, and adding support for `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu` vfixed-point instructions. What we have done so far is as follows:
> > 
> > * We have added support for `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu` vfixed-point instruction and have tested them by using the provided tests
> > * We have fixed the RTL failures of `vasub`, `vasubu` and `vssub` vfixed-point instructions
> > * We have fixed CI check failures of `vsadd`, `vsaddu` and `vssub` instructions
> > * We have fixed `vssub.c` test which caused the spike failure
> > 
> > The final testing of all the bug-fixes and RTL additions is on its way. We are planning to push the fixes as soon as we get the final results.
> > Thanks
> 
> That is great! are your changes based of this PR?

Yes

**Comment by M-Ijaz-10x** [comment]
Hey @hossein1387

As our changes are based of this PR, should we add our changes to this PR or create a new PR (Because there are a few major additions and it would be cleaner to create a new PR)?

Thanks!

**Comment by hossein1387** [comment]
> Hey @hossein1387
> 
> As our changes are based of this PR, should we add our changes to this PR or create a new PR (Because there are a few major additions and it would be cleaner to create a new PR)?
> 
> Thanks!
 @mp-17 @suehtamacv  what do you think?
I am actually working on this issue now but I do not have `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu`.

**Comment by M-Ijaz-10x** [comment]
Hey @hossein1387 

Its great that you are already working on resolving the CI check failures issue, we can wait for this PR to be merged into `pulp-platform/ARA/main` and later on we can add support for `vsmul`, `vssrl`, `vssra`, `vnclip` and `vnclipu` vfixed-point instructions in separate PR.

Thanks!

**Comment by suehtamacv** [comment]
> Hey @hossein1387
> 
> As our changes are based of this PR, should we add our changes to this PR or create a new PR (Because there are a few major additions and it would be cleaner to create a new PR)?
> 
> Thanks!

You could add a new PR whose base is this branch. That would be clean, and then we could merge the two PRs one after the other

**Comment by mp-17** [comment]
Thank you @M-Ijaz-10x @hossein1387 for the joint effort!
In parallel, I am solving the backend problems we meet after synthesis. I would freeze the merge of this PR until then

**Comment by mp-17** [comment]
We can close timing after place & route, I will check again before merging, but everything seems okay! ;-)

**Comment by mp-17** [comment]
@hossein1387 , I have created a PR on your GitHub to fix the problems with the CI https://github.com/PolyMTL-Gr2m/ara/pull/1

**Comment by mp-17** [comment]
Switching to: https://github.com/pulp-platform/ara/pull/147
Thanks Hossein!

**Pull request closed without merging**
